osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -443,29 +445,43 @@ public void Filter(FilterCriteria newCriteria, bool debounce = true)
 
         private void applyActiveCriteria(bool debounce, bool alwaysResetScrollPosition = true)
         {
-            PendingFilter?.Cancel();
-            PendingFilter = null;
+            NeedsFiltering = true;
+            if (debounce && (FilteringCooldown != null)) {
+                // The `FilteringCooldown` task executes as soon as filtering
+                // is off cooldown. When it executes, it sees that `NeedsFiltering`
+                // is set to true and calls `perform`. This will happen no more
+                // than 100 milliseconds from now.
+                return;
+            }
 
-            if (debounce)
-                PendingFilter = Scheduler.AddDelayed(perform, 250);
+            // If initial load is not yet finished, this will be run inline in
+            // loadBeatmapSets to ensure correct order of operation.
+            if (!(BeatmapSetsLoaded || debounce))
+                PendingFilter = Schedule(perform);
             else
-            {
-                // if initial load is not yet finished, this will be run inline in loadBeatmapSets to ensure correct order of operation.
-                if (!BeatmapSetsLoaded)
-                    PendingFilter = Schedule(perform);
-                else
-                    perform();
-            }
+                perform();
 
             void perform()
             {
                 PendingFilter = null;
+                NeedsFiltering = false;
 
                 root.Filter(activeCriteria);
                 itemsCache.Invalidate();
 
                 if (alwaysResetScrollPosition || !Scroll.UserScrolling)
                     ScrollToSelected(true);
+
+                FilteringCooldown = Scheduler.AddDelayed(cooldown, 100);
```

Comment:
Have you tested this with 200,000 panels? Such testability already exists in visual tests and I would hope you have. But based on experience it doesn't seem like you have as I think this will considerably impact performance with the current search logic. There's a reason we chose the numbers that are currently present.

Also a reason we use debounce in its actual sense rather than your "run-every-interval" implementation.

Reply:
Focusing on the test failure is missing the point. Benchmarking performance is the point and that won't usually fail tests (except in the most extreme cases). You should be either running benchmarks, profiling, or at the very least opening the debug performance display (<kbd>Ctrl</kbd>+<kbd>F2</kbd>) and visually roughly cross-checking for regressions wrt the registered `Update()` runtime of the carousel components therein.

Reply:
You can also just (temporarily) delete the failing asserts and test interactively.

Reply:
> You can also just (temporarily) delete the failing asserts and test interactively.

That was my first instinct, but the BeatmapCarousel visual test environment doesn't support search/filtering using keyboard input.

Reply:
You should be adding steps to perform the testing, not manually testing.